File: code/0001. Two Sum/solution3.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        int n = nums.length;

        // create sorted array
        List<Pair<Integer, Integer>> sortedWithIndex = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            sortedWithIndex.add(new Pair<>(nums[i], i));
        }
        Collections.sort(sortedWithIndex, Comparator.comparing(Pair::getKey));

        // two pointers
        int p1 = 0, p2 = n - 1;
        while (p1 < p2) {
            int num1 = sortedWithIndex.get(p1).getKey(), num2 = sortedWithIndex.get(p2).getKey();
            if (num1 + num2 == target) {
                return new int[] {sortedWithIndex.get(p1).getValue(), sortedWithIndex.get(p2).getValue()};
            }
            else if (num1 + num2 < target) {
                p1++;
            }
            else {
                p2--;
            }
        }

        return null;
    }
}

File: code/0015. 3Sum/solution3.cpp
class Solution {
private:    
    void twoSum(vector<int>& nums, int target, int start, vector<vector<int>>& result) {
        int p1 = start, p2 = nums.size() - 1;
        while (p1 < p2) {
            int num1 = nums[p1], num2 = nums[p2];
            if (num1 + num2 == -target) {
                result.push_back({target, num1, num2});
                if (num1 == num2) {
                    return;
                }
                while (nums[p1] == num1) {
                    p1++;
                }
                while (nums[p2] == num2) {
                    p2--;
                }
            }
            else if (num1 + num2 < -target) {
                p1++;
            }
            else {
                p2--;
            }
        }
    }
    
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;
        
        for (int i = 0; i < n; i++) {
            if (i == 0 || nums[i] != nums[i - 1]) {
                int target = nums[i];
                twoSum(nums, target, i + 1, result);
            }
        }
        
        return result;
    }
};